### PR TITLE
Dynamic available help channels message

### LINKED ISSUE #1320: Add a dynamic message at the bottom of `#how-to-get-help`
Actions https://github.com/python-discord/organisation/issues/315. 

We should have a dynamic message at the bottom of #how-to-get-help saying something along the lines of 
“The currently available channels are #help-element1 and #help-element2, feel free to ask in any of those two channels!”, 
to avoid getting anyone getting confused by the currently open channels like we had in the past.

### PR DESCRIPTION
**Pull request by Xithrius**
Closes #1320
Rewrite of #1325

Users sometimes get confused on which channels are available for them to claim.

With this PR, a message in the #how-to-get-help channel will be edited to reflect which channels are currently available.

If the bot cannot edit the most recent message in the #how-to-get-help channel, then it knows it's trying to edit the webhook message that contains the guide for claiming a channel. It will send the dynamic message if this occurs.

If we do decide to remove or add more channels in the `help: available` category, then the bot will be able to pick up on them.

image

### CODE REVIEW

File: bot/exts/help_channels/_cog.py [hunk 1]
```diff
@@ -461,3 +487,25 @@ async def wait_for_dormant_channel(self) -> discord.TextChannel:
...
+            self.available_help_channels = set(
+                c.id for c in available_channels_category.channels if 'help-' in c.name
+            )
+
+        available_channels = AVAILABLE_HELP_CHANNELS.format(
+            available=', '.join(f"<#{c}>" for c in self.available_help_channels) or None
+        )
+
+        if self.how_to_get_help is None:
+            self.how_to_get_help = await channel_utils.try_get_channel(Channels.how_to_get_help)
```
Comment (HassanAbouelela) [inline bot/exts/help_channels/_cog.py:504]: Could you add a check for `how_to_get_help` being an instance of TextChannel? GuildChannels could be voice or category channels, which don't all impleemnt fetch_message, send, and last_message.

Perhaps we should log a warning and abort if the channel is not of the correct type.

File: bot/exts/help_channels/_cog.py [hunk 2]
```diff
@@ -332,6 +354,10 @@ async def move_to_available(self) -> None:
             category_id=constants.Categories.help_available,
         )
 
+        # Adding the help channel to the dynamic message, and editing/sending that message.
+        self.available_help_channels.add(channel.id)
+        await self.update_available_help_channels()
```
Comment (HassanAbouelela) [inline bot/exts/help_channels/_cog.py:359]: You can save one edit if you remove this update. The next thing that runs after this task is another update.

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]
Everything is looking solid so far. I have a few functionality concerns, but everything works as expected in general.

File: bot/exts/help_channels/_cog.py (hunk 1 above)
Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:504]: I don't think this is necessary since there's no plans of having a voice channel anywhere near the available help channels.

File: bot/exts/help_channels/_cog.py [hunk 3]
```diff
@@ -461,3 +486,25 @@ async def wait_for_dormant_channel(self) -> discord.TextChannel:
         self.queue_tasks.remove(task)
 
         return channel
+
+    async def update_available_help_channels(self) -> None:
+        """Updates the dynamic message within #how-to-get-help for available help channels."""
+        if not self.available_help_channels:
+            available_channels_category = await channel_utils.try_get_channel(Categories.help_available)
```
[the remark this answers is not in the document]
Comment (Xithrius) [reply bot/exts/help_channels/_cog.py]: Forgot that the `self.available_category` attribute exists.

**Review by HassanAbouelela — APPROVED** [review]
Clean, and works as expected. LGTM

File: bot/exts/help_channels/_cog.py [hunk 4]
```diff
@@ -20,6 +21,9 @@
 HELP_CHANNEL_TOPIC = """
 This is a Python help channel. You can claim your own help channel in the Python Help: Available category.
 """
+AVAILABLE_HELP_CHANNELS = """
+Currently available help channel(s): {available}
```
Comment (Shivansh-007) [inline bot/exts/help_channels/_cog.py:25]: How does it look if we do this:
```suggestion
**Currently available help channel(s)**: {available}
```

**Review by Shivansh-007 — APPROVED** [review]
Looks Good!

File: bot/exts/help_channels/_cog.py (hunk 4 above)
Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:25]: image

I prefer the first one.

Comment (Shivansh-007) [reply bot/exts/help_channels/_cog.py:25]: Hmm, i like it with bold, let's see what others prefer.

Comment (Vthechamp22) [reply bot/exts/help_channels/_cog.py:25]: I think it looks fine normally (without bold).

Comment (Akarys42) [reply bot/exts/help_channels/_cog.py:25]: I prefer the bold version to catch the eyes

Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:25]: After further consideration I see the appeal of the bold. I will change this now.

**Comment by Xithrius** [comment]
Here's what the dynamic message now looks like.
image

File: bot/exts/help_channels/_cog.py [hunk 5]
```diff
@@ -114,6 +123,9 @@ async def claim_channel(self, message: discord.Message) -> None:
 
         await _caches.unanswered.set(message.channel.id, True)
 
+        # Removing the help channel from the dynamic message, and editing/sending that message.
+        self.available_help_channels.remove(message.channel.id)
```
Comment (Numerlor) [inline bot/exts/help_channels/_cog.py:127]: Why are we storing ids instead of the channels themselves?

File: bot/exts/help_channels/_cog.py [hunk 6]
```diff
@@ -20,6 +21,9 @@
 HELP_CHANNEL_TOPIC = """
 This is a Python help channel. You can claim your own help channel in the Python Help: Available category.
 """
+AVAILABLE_HELP_CHANNELS = """
+**Currently available help channel(s):** {available}
+"""
```
Comment (Numerlor) [inline bot/exts/help_channels/_cog.py:26]: This fits into one line comfortably

File: bot/exts/help_channels/_cog.py [hunk 7]
```diff
@@ -275,6 +287,15 @@ async def init_cog(self) -> None:
...
         self.close_command.enabled = True
 
+        # Getting channels that need to be included in the dynamic message.
+        task = asyncio.create_task(self.update_available_help_channels())
+        self.queue_tasks.append(task)
+
+        await task
+
+        log.trace("Dynamic available help message updated.")
+        self.queue_tasks.remove(task)
```
Comment (Numerlor) [inline bot/exts/help_channels/_cog.py:297]: What's the reason for creating a task and immediately awaiting it? If it's because of `queue_tasks` and their cancellation then it should be unnecessary, as the init task itself is cancelled when the cog is unloaded

File: bot/exts/help_channels/_cog.py [hunk 8]
```diff
@@ -461,3 +486,24 @@ async def wait_for_dormant_channel(self) -> discord.TextChannel:
         self.queue_tasks.remove(task)
 
         return channel
+
+    async def update_available_help_channels(self) -> None:
+        """Updates the dynamic message within #how-to-get-help for available help channels."""
+        if not self.available_help_channels:
+            self.available_help_channels = set(
+                c.id for c in self.available_category.channels if 'help-' in c.name
```
Comment (Numerlor) [inline bot/exts/help_channels/_cog.py:494]: Use the `_channel.is_excluded_channel` util

**Review by Numerlor — CHANGES_REQUESTED** [review]

File: bot/exts/help_channels/_cog.py (hunk 5 above)
Comment (mbaruh) [reply bot/exts/help_channels/_cog.py:127]: +1 to this. It will also make formatting the message more straightforward with channel.metion

File: bot/exts/help_channels/_cog.py (hunk 7 above)
Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:297]: I thought back then that it would be necessary for this task to be cancelled if the cog was unloaded. 

I now see that there's no significant reason to do so.

File: bot/exts/help_channels/_cog.py [hunk 9]
```diff
@@ -20,6 +21,7 @@
 HELP_CHANNEL_TOPIC = """
 This is a Python help channel. You can claim your own help channel in the Python Help: Available category.
 """
+AVAILABLE_HELP_CHANNELS = """**Currently available help channel(s):** {available}"""
```
Comment (Shivansh-007) [inline bot/exts/help_channels/_cog.py:24]: Can we spread this across lines?
```suggestion
AVAILABLE_HELP_CHANNELS = """
**Currently available help channel(s):** {available}
"""
```

File: bot/exts/help_channels/_cog.py [hunk 10]
```diff
@@ -461,3 +479,24 @@ async def wait_for_dormant_channel(self) -> discord.TextChannel:
...
+
+        if self.how_to_get_help is None:
+            self.how_to_get_help = await channel_utils.try_get_channel(Channels.how_to_get_help)
+
+        try:
+            if self.dynamic_message is None:
+                self.dynamic_message = await self.how_to_get_help.fetch_message(self.how_to_get_help.last_message_id)
+            await self.dynamic_message.edit(content=available_channels)
+        except discord.NotFound:
+            self.dynamic_message = await self.how_to_get_help.send(available_channels)
```
Comment (Shivansh-007) [inline bot/exts/help_channels/_cog.py:502]: Can we have a `log.trace()` statement here?

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]

File: bot/exts/help_channels/_cog.py (hunk 9 above)
Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:24]: https://github.com/python-discord/bot/pull/1396#discussion_r570897703

Comment (Shivansh-007) [reply bot/exts/help_channels/_cog.py:24]: It seems more readable, `"""` are multi line strings, so either change it to `"` single quotes or keep it as it was before
(my opinion)

Comment (Xithrius) [reply bot/exts/help_channels/_cog.py:24]: After looking at other cogs, it does seem that one liners should have single double quotes. An example can be seen [here](<https://github.com/python-discord/bot/blob/e5b073c5ee9c7c887a193ec05d813d259eca58ee/bot/exts/help_channels/_message.py#L30>).

**Review by Shivansh-007 — APPROVED** [review]
Thanks! Looks Good To Me!

**Review by HassanAbouelela — APPROVED** [review]
Anyone using the server template will need to update their config to move the how-to-get-help channel out of the help available category, but it shouldn't be a big deal. This is not important for the test server, as it uses the same channel for cooldown and how-to-get-help.

**Review by gustavwilliam — APPROVED** [review]
Thanks! Works like a charm — unless we go into some edge cases.

When lowering the maximim amount of help channels (restarting bot in-between), the higher number of help channels will be listed in the available help channels message. Let me show you an example, where we go from 3 to 2 channels. After restarting the bot with the new config, the following will happen:

- The bot keeps updating the message with 3 help channels. One of them will be there statically (and incorrectly), while the other two are correctly and continuously updated
- When deleting the original "these channels are available" message, the bot will resend the same three-channel message

One workaround for this is deleting the message and *then* restarting the bot.

The other edge case comes into play when we increase the number of help channels that should be available, like going from 2 to 3. This simply sends a second message with two channels and keeps updating that. Just delete the first message manually and things will work just fine.

However, both of these scenarios are uncommon enough that I'll still approve this PR. If you want to fix the mentioned bugs, that would be fantastic, but I really don't mind leaving it at this point. It works flawlessly, unless we get into some uncommon and strange scenarios.

**Comment by Xithrius** [comment]
Thank you for your approval! 

I think that any of these edge cases will have to include a bot redeploy, as rules would probably have to be changed in order for them to take effect. If we encounter these edge cases, I'll be sure to put it into high/critical priority.

**Comment by Xithrius** [comment]
Bot can fetch a message that was previously deleted, which causes a `discord.NotFound` error since it cannot find the message to edit.

On the actual server, no message has been deleted. The bot this time finds the webhook message, then tries to edit it.

I will make a change to check if the last message was sent by the bot. If not, it will send a dynamic message (or something similar to this)

image

**Comment by jb3** [comment]
Where are we at on this? Should we open a new PR?

**Comment by Xithrius** [comment]
I will be opening a new PR for this, I'm still working out the logic.
